Feature/ppm 5 read clients persistent db at controller init

adam1985d opened the pull request:
This PR introduces the flow of reading clients from the persistent DB as part of the controller init flow.

As part of this PR the following improvements/changes were introduced:
- persistent DB param-names are public to `db` class users (required for efficient update of the multiple parameters at once).
- In existing db APIs verify persistent-db is enabled before accessing the BPL.
- missing function description was added.
- added a client-aging and removal functionality (both for runtime DB parameters and for persistent DB).
- added a counter of clients in persistent-db and moved all BPL add/remove client operation to use private `db` class APIs that encapsulates the counter update.
- added a "remove-a-client-before-add" functionality if persistent DB is full.
- refactored the load_persistent_db_clients() to handle the filtering of aged and invalid client entries.
- added a call to load_persistent_db_clients() from the controller init flow - if the persistent DB is enabled.

relates to PPM-5

adam1985d committed:
controller: db: make persistent params param names public

PPM-5.
Preparative Commit.
Add header file public forward declaration of persistent db param names
to allow controller-db to construct a one-shot set/add of a client with
a map of key-value of params.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: check persistent db enabled before bpl access

PPM-5.
Added check of persistent_db value in the controller config before
accessing the BPL.
For relevant APIs, return an error if called with persistent-db is disabled.
For relevant APIs, skip BPL-access and perform only runtime-db
configuration.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add missing function description

PPM-5.
Added missing function description to update_client_entry_in_persistent_db().

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add persistent db clients counter

PPM-5.
Preparative commit.
Added persistent db clients counter to the db class.
Added private functions to add/remove clients that encapsulate the
increment/decrement of the counter.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: replace bpl direct add/remove with db APIs

PPM-5.
Replaced direct BPL calls to add/remove a client from to/from the
persistent db with calls to the db class APIs which encapsulate the
increment/decrement of the persistent db clients' counter respectively.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add API to set node params from map

PPM-5.
Added API to set node-params from a map of <param-name,param-value>.
This API enables us to update multiple params in a node in a single-shot
which provides efficiency compared to performing a separate call update
using the existing db class APIs.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add function to get client to be removed

PPM-5.
Added function to get the preferred client to be removed from the
persistent-db.
The preference mentioned above is according to the following, and is
detailed in the function description:
- prefer disconnected clients over connected clients.
- according to the above, find the client with the least time left
before aging.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add function to remove most aged client

PPM-5.
Added function to remove the client provided by the
get_candidate_client_for_removal() function.
The function uses the clear_client_persistent_db() which clears any
persistent parameters from the runtime DB and removes the client from
the persistent DB.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: remove a client if persistent db is full

PPM-5.
Added a check of the clients counter to the
add_client_to_persistent_db() function. If persistent DB is full, a
client is removed before adding the new client to the persistent DB.
The removal of a client utilizes the remove_candidate_client() function.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: refactor load_persistent_db_clients()

PPM-5.
The load_persistent_db_clients() API current implementation, and purpose
is to provide an interface to read all the clients from the BPL, and the
responsibility of adding nodes for those clients as well as filter-out
aged clients was on the caller.
A more suitable approach is to handle all of the above internally and
leave the caller of the API only the responsibility of triggering it.

Changed the API return value to success/fail instead of a map of
clients.
Filtered-out aged client-entries and client-entries not containing a
valid timestamp. The filtered-out clients are removed from persistent
DB as well as not added to runtime DB.
Added nodes for "valid" client entries with the persistent data read
from persistent DB.

While at it, added an is_valid_mac() check to the
client_db_entry_to_mac() implementation.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: son_master: load clients from persistent db

PPM-5.
Added a call to the load_persistent_db_clients() API, if the persistent
DB is enabled.
The call is performed at the beginning of the master_thread::init() to
make sure the clients persistent information is loaded before any task
is started.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2484,6 +2484,11 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    }
     auto client_db_entry = client_db_entry_from_mac(mac);
 
     return bpl::db_has_entry(type_to_string(beerocks::eType::TYPE_CLIENT), client_db_entry);
```
This is confusing.
The last return statement says that the return value could be false if the given mac is not on the database, whereas on the if condition the return of false means that the function has failed.

adam1985d replied:
you are correct - removed this check - it is not appropriate to this function

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2542,16 +2553,21 @@ bool db::set_client_time_life_delay(const sMacAddr &mac,
+            LOG(ERROR) << "persistent db is disabled";
+        } else {
+            LOG(DEBUG) << "configuring persistent-db, timelife = " << time_life_delay_sec.count();
 
```
Why don't you return `false` here? (I see it on few places on this file).
And if you did:
```cpp
if (error) {
   // handle error
}
// rest of the code
```

adam1985d replied:
this is since the parameters can be utilized even if `persistent-db=disable` mode by the controller tasks or by the user, the log print is to indicate the calls to the API with `save_to_persistent_db=true` will save the information to run-time db but are not persistent since the persistent-db is disabled

morantr replied:
In that case, it is not an error, and you should use DEBUG/WARNING/INFO  (probably debug) on the log print.
I probably would pass on this debug print and do it like this:
```cpp
if (save_to_persistent_db && config.persistent_db) {
   // code
}
```

adam1985d replied:
but then you don't reflect the "misuse" to the user (developer)

adam1985d replied:
changed to a DEBUG print as we agreed upon

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2542,16 +2553,21 @@ bool db::set_client_time_life_delay(const sMacAddr &mac,
+        if (!config.persistent_db) {
+            LOG(ERROR) << "persistent db is disabled";
+        } else {
+            LOG(DEBUG) << "configuring persistent-db, timelife = " << time_life_delay_sec.count();
```
Now that I look at it, "timelife" looks very strange to me. I don't think it is even a valid word combination in English.
Maybe it should be "lifetime" or "time **of** life".

adam1985d replied:
I actually felt the same on first encounter - but this is the term used repeatedly in the feature requirements

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2585,16 +2601,22 @@ bool db::set_client_stay_on_initial_radio(const sMacAddr &mac, bool stay_on_init
-        values_map.at(initial_radio_enable_str) = std::to_string(stay_on_initial_radio);
+            std::unordered_map<std::string, std::string> values_map;
+            values_map.at(timestamp_str)            = timestamp_to_string_seconds(timestamp);
+            values_map.at(initial_radio_enable_str) = std::to_string(stay_on_initial_radio);
```
You can't use `.at()`. To use `.at()` you need to be sure that the key exists on the map. If it doesn't exist, an exception will be thrown and the process will crash. Since you declare `values_map` here, it definitely does not exist.

You did it on other places in this file as well.

adam1985d replied:
you are right, but many of this changes were introduced by the previous PR.

added a commit that replaces all insertions that use `.at()` with use of `operator[]`.

all new commits were also aligned accordingly.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2585,16 +2601,22 @@ bool db::set_client_stay_on_initial_radio(const sMacAddr &mac, bool stay_on_init
 
     auto timestamp = std::chrono::steady_clock::now();
```
It is probably best if you declare it where you need it. Otherwise, it could be recalled and not being used (in case of failure).

adam1985d replied:
in the general case this is true - but here I use it to define the client last-edit timestamp, which I wish to be the same in the persistent-db and runtime-db - so value needs to be determined in common scope

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4233,3 +4233,28 @@ bool db::update_client_entry_in_persistent_db(
     return true;
 }
+
+bool db::add_client_entry_and_update_counter(
```
I don't think you should mention the "update_counter" explicitly in the function name, it is already on the function documentation.
This is just my opinion, feel free to ignore it.

adam1985d replied:
I think this should be emphasized to prevent any new implementation using this API to externally update it and cause mayhem.

morantr replied:
I don't think it is a reasonable consideration. As a user, I would like to know less as I can about the internal mechanism.

adam1985d replied:
this is a private function

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1105,6 +1105,25 @@ class db {
                                          std::unordered_map<std::string, std::string> values_map);
 
+    /**
+     * @brief adds a client entry to persistent_db with configured parameters and increments clients counter.
```
Missed the capital letter at sentence beginning.

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2505,9 +2505,9 @@ bool db::add_client_to_persistent_db(const sMacAddr &mac,
 
-    if (!bpl::db_has_entry(std::string("type"), db_entry)) {
+    if (!bpl::db_has_entry(type_to_string(beerocks::eType::TYPE_CLIENT), db_entry)) {
         // if entry already exists in DB
```
This comment confuses me. You check it DB does not have an entry, but you write that entry already exists.
Same on line 2561.

adam1985d replied:
you are correct but vise-versa - the condition should not have been with `!` - fixed it for both occurrences.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -20,12 +20,12 @@ using namespace beerocks_message;
+const std::string db::initial_radio_enable_str = "initial_radio_enable";
+const std::string db::initial_radio_str        = "initial_radio";
+const std::string db::selected_band_enable_str = "selected_band_enable";
+const std::string db::selected_bands_str       = "selected_bands";
```
I don't understand why you changed it to be class members instead of global const on the cpp file.
Also, for constants please use CAPITAL_LETTERS_SNAKE_CASE, this way it is easy to now that it is constant.

adam1985d replied:
they are part of the class as they are only relevant to `db` class users - I know I can just as much define them globally in the header - but didn't for that reason.

will CAPITALIZE them

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4309,3 +4309,64 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
+    const auto max_timelife_delay_sec =
+        std::chrono::seconds(config.max_timelife_delay_days * 24 * 3600);
+    //const auto unfriendly_max_timelife_delay_sec =
+    //    std::chrono::seconds(config.unfriendly_device_max_timelife_delay_days * 24 * 3600);
```
What is this comment for?

adam1985d replied:
I will remove it - we decided to move this functionality to future PR

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1133,6 +1133,16 @@ class db {
+     *
+     * @return sMacAddr mac of candidate client to be removed - if not found, string_utils::ZERO_MAC is returned.
+     */
+    sMacAddr get_next_to_be_aged_client_preferably_disconnected();
```
Damn! It is a long function name! 🙂
Does this function purpose is to find a client to remove because the database is full-up or something?
I think you should rename it to "get_removal_candidate_client()"

The full explanation about how the function does what it does should be on the documentation, not in the function name.

adam1985d replied:
I was hoping for naming suggestions on this one - maybe `get_candidate_client_for_removal()` would be better?

morantr replied:
Sounds good to me. 🙂

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4309,3 +4309,64 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
+            if (client->get_type() == beerocks::eType::TYPE_CLIENT) {
+                //TODO: improvement - stop search if "already-aged" candidate is found (don't-care of connectivity status)
+
+                // skip clients which has no persisten information
```
```suggestion
                // Skip clients which have no persistent information.
```

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4309,3 +4309,64 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
+                    continue;
+                }
+
+                // prefering disconnected clients over connected ones (even if less aged)
```
```suggestion
                // Preferring disconnected clients over connected ones (even if less aged).
```

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4309,3 +4309,64 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
+
+                // client timelife delay
+                //TODO: use max_timelife_delay_sec/unfriendly_max_timelife_delay_sec according to "is_unfriendly" check which is yet-to-be-defined
+                auto lifetime_delay =
```
Here you wrote "lifetime" (which is a valid word), but in other places, you wrote "timelife" (which is not a valid word).
Need to use the same term everywhere.

adam1985d replied:
you are correct - fixed to use `timelife_delay`

morantr replied:
I kind of wanted you will change it to "lifetime" in other places... 😕

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4309,3 +4309,64 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
+                // compare to currently chosen candidate expiry due time
+                if (current_client_expiry_due < candidate_client_expiry_due_time) {
+                    candidate_client_expiry_due_time = current_client_expiry_due;
+                    if (client->state == beerocks::STATE_DISCONNECTED) {
+                        is_disconnected_candidate_available = true;
+                    }
+                    candidate_client_to_be_removed = tlvf::mac_from_string(key_value.first);
```
If on the next iteration `current_client_expiry_due < candidate_client_expiry_due_time` is `true`, but `client->state == beerocks::STATE_DISCONNECTED` is `false` and on one of the previous iterations `is_disconnected_candidate_available`  has been marked as `true`, the `candidate_client_to_be_removed` will be a connected client and `is_disconnected_candidate_available` will be true although it is not.
You need to decide what condition is stronger.
If it is the connection status (prefer disconnect client) it should look like this:
```cpp
if (current_client_expiry_due < candidate_client_expiry_due_time) {
    candidate_client_expiry_due_time = current_client_expiry_due;
    if (!is_disconnected_candidate_available  && client->state == beerocks::STATE_DISCONNECTED) {
        is_disconnected_candidate_available = true;
    } else if (is_disconnected_candidate_available && client->state == beerocks::STATE_CONNECTED) {
        // Skip client because it is connected and we already have disconnected client.
        continue;
    }
    candidate_client_to_be_removed = tlvf::mac_from_string(key_value.first);
}
```

If the client_expiry_due_time is stronger, it should look like this:
```cpp
if (current_client_expiry_due < candidate_client_expiry_due_time) {
    candidate_client_expiry_due_time = current_client_expiry_due;
    is_disconnected_candidate_available = client->state == beerocks::STATE_DISCONNECTED;
    candidate_client_to_be_removed = tlvf::mac_from_string(key_value.first);
}
```

adam1985d replied:
the handling of the case where a disconnected candidate client was found and the current client is connected is already handled above by:
```
                 if (client->client_parameters_last_edit ==
                    std::chrono::steady_clock::time_point::min()) {
                    continue;
                }
```

but will add a comment to make it more clear

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1133,6 +1133,13 @@ class db {
+     * 
+     * @return true on success, otherwise false.
+     */
+    bool remove_next_to_be_aged_client_preferably_disconnected();
```
Here also, the function name is too long.
I would change it to "remove_next_client()".

adam1985d replied:
accordingly probably if the `get_candidate_client_for_removal()` is fine by you in previous commit then this would change to `remove_candidate_client()`

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+
+    //validate clients data from persistent db
+    for (auto &client : clients) {
+        const auto &client_entry = client.first;
```
What is in the entry? I see it is a string. Is it a mac? If it does, maybe it will be best to explicitly mention it.

adam1985d replied:
we cannot use the mac as-is since the UCI (used by the BPL) doesn't permit the `:` in the section-name
for that purpose, we replace the `:` with `_`

I wanted to differ here from syntax used for db-entries and the mac itself - don't want people to mistakenly use the `client.first` as a mac without transforming it first

morantr replied:
So it is a mac, only formatted like this? `aa_bb_cc_dd_ee_ff` ?
Please check is it possible to use colons on the UCI if the phrase it wrapped with double or single quotes or something. You might as well use hex format, which is wildly used: `0xaabbccddeeff` and you could convert it with to sMacAddr with utility functions we already have.
Anyway, I don't think it is a good idea to just leave it like this, this is confusing already and will look ugly/strange on the log.
I think that on the load function you should convert whatever format you use on UCI, to `sMacAddr`,
and then of course, rename `client_entry` to `client__mac`.

adam1985d replied:
again - please differ between BPL-context - which is referred to as entry - to the DB/nodes context which we refer to as client-mac
and to your question - no - the UCI doesn't support the colons in any form (as section name, as the value of param, aka. option, it does)

I intentionally used the entry and not the transformation from entry to mac to allow easy debugging in anyone edited the UCI and didn't provide an appropriate entry name - you will easily find the "culprit" since you have the invalid entry-name.

this actually raises another case where the entry name is invalid - I will add this verification as part of the refactoring of this API implementation.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+        if (timestamp_it == client_data_map.end()) {
+            LOG(ERROR)
+                << "client persistent data without timestamp is invalid - setting client entry "
+                << client_entry << "as invalid";
```
```suggestion
                << client_entry << " as invalid";
```

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+            //        check against unfriendly_device_max_timelife_delay_days
+            if (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::steady_clock::now() -
+                                                                 timestamp)
+                    .count() > max_timelife_delay_sec) {
```
It would look better if you do the calculation outside the `if`, and put on the `if` a variable containing the result.

adam1985d replied:
done

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+        } else {
+            ++it;
+        }
+    }
```
Why to iterate over `clients` three times? On the first loop you know already that a client is invalid, so remove it there from the list and from the db, instead of marking it as "invalid" and loop over it, again and again, to remove from the map and from the DB.

adam1985d replied:
I thought it would be more readable to handle each check separately:
1. check for invalid clients
2. check for aged clients
3. remove clients from persistent-db
4. add "valid" clients to nodes-map and update their persistent data

I will refactor this function implementation

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+                << "client persistent data without timestamp is invalid - setting client entry "
+                << client_entry << "as invalid";
+            //set client as invalid - all invalid clients will be removed from map before adding nodes for clients
+            client_data_map.at(timestamp_str) = invalid_timestamp;
```
You can't use ".at()" here, the key does not exist. It will throw an exception and the process will crash.

adam1985d replied:
went over all relevant locations and replaced with `operator[]` after your first comment.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+                << client_entry << "as invalid";
+            //set client as invalid - all invalid clients will be removed from map before adding nodes for clients
+            client_data_map.at(timestamp_str) = invalid_timestamp;
+        } else {
```
Please use:
```cpp
if (error) {
   // handle error
}
// rest of the code

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2904,27 +2904,116 @@ bool db::update_client_persistent_db(const sMacAddr &mac)
+            continue;
+        }
+        ++clients_added_no_error;
     }
```
This is something you can do as well on the first loop in the function.

adam1985d replied:
will refactor this function entirely

morantr commented on controller/src/beerocks/master/beerocks_master_main.cpp:
```diff
@@ -344,6 +344,17 @@ int main(int argc, char *argv[])
+        } else {
+            LOG(DEBUG) << "load clients from persistent db finished successfully";
+        }
+    }
```
Why is it here and not on `son_master_thread::init()`?

adam1985d replied:
same as why the `master_db` is created before the `son_master_thread:init()` - I consider this operation as pre-init preparations.

morantr replied:
The database was created here originally because it was planned to be a shared object between several threads on the process, not because it is a "pre-init" operation.
That's why the database lock/unlock feature was implemented.
I believe that the `main` files should be with less code as possible, and if there is not a really good reason to put it here, then it should be on the `son_master::init()`

adam1985d replied:
I will move in inside the `son_master::init()`

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2495,7 +2495,7 @@ bool db::is_client_in_persistent_db(const sMacAddr &mac)
 
 bool db::add_client_to_persistent_db(const sMacAddr &mac,
-                                     std::unordered_map<std::string, std::string> params)
+                                     const std::unordered_map<std::string, std::string> &params)
```
Please don't add a separate commit to fix Cppcheck issues.
Fix it directly on the commits you added the code which raised the issues.

adam1985d replied:
fixed

morantr reviewed (changes requested):
Request changes since I noticed issues that may cause a crash.

adam1985d committed:
controller: db: replace .at() with operator[] in unordered map

The .at() function of unordered-map should be used for const maps and
for items in the map we are certain exist and not for addition of new
items where operator[].

Replaced all occurences of .at() with use of operator[].

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d reviewed (commented):
@morantr please go over my responses and give your replies - thanks ahead

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -53,6 +53,11 @@ std::string db::type_to_string(beerocks::eType type)
+    if (!network_utils::is_valid_mac(db_entry)) {
+        return {};
+    }
+
```
That is redundant. `network_utils::is_valid_mac` will work only if the given mac is a string.
If the string has an illegal character like "AA:BB:CC:DD:EE:**ZZ**" the validation function will return false.
In the case of `sMacAddr` there is nothing to validate. Since `sMacAddr` is just 6 octets, it is promised to be valid.

adam1985d replied:
removed validation

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -64,6 +69,10 @@ sMacAddr db::client_db_entry_to_mac(const std::string &db_entry)
 
     std::replace(entry.begin(), entry.end(), '_', ':');
```
Consider changing the input argument to not be a const, if it is possible. If not possible, just don't use reference.

adam1985d replied:
will be fixed in the next PR (this function is not added in this PR and this change - although I agree with it - should not be part of this PR)

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -187,14 +187,15 @@ class db {
      * 
      * @param mac MAC address of a client.
      * @return std::string the string representation of the MAC address with ':' replaced with '_' removed.
+     * @return An empty string is returned on failure.
```
And on success?

adam1985d replied:
one line above it :-)

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -791,12 +792,12 @@ class db {
 
     /**
      * @brief Load all clients from persistent db.
+     * Creates nodes for the clients in runtime-db and set persistent parameters values accordingly
```
Missing a dot at the end of the line.

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/beerocks_master_main.cpp:
```diff
@@ -344,6 +344,7 @@ int main(int argc, char *argv[])
     son::db master_db(master_conf, logger, bridge_info.mac);
     LOG(DEBUG) << "slave_keep_alive_retries=" << master_db.config.slave_keep_alive_retries;
     // diagnostics_thread diagnostics(master_db);
+
```
Please remove this space.

adam1985d replied:
removed it

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -88,6 +88,16 @@ bool master_thread::init()
+    if (database.config.persistent_db) {
+        LOG(DEBUG) << "loading clients from persistent db";
+        if (!database.load_persistent_db_clients()) {
+            LOG(ERROR) << "failed to load clients from persistent db";
```
If it is an ERROR, you should return false. If it is not critical and you wish to continue on that scenario, print a WARNING.

morantr replied:
Also, the headline of the commit message is referring to "master_main" which is not (anymore).

adam1985d replied:
changed the ERROR log print to WARNING

adam1985d replied:
fixed the commit message

morantr reviewed (approved):
Nothing critical, so approving.

RanRegev commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -20,12 +20,12 @@ using namespace beerocks_message;
-static const std::string selected_bands_str       = "selected_bands";
+const std::string db::TIMESTAMP_STR            = "timestamp";
+const std::string db::TIMELIFE_DELAY_STR       = "timelife";
+const std::string db::INITIAL_RADIO_ENABLE_STR = "initial_radio_enable";
```
nitpick:
string are strings: underscore is not different from space in any aspect, so why go "technical", why not simply use natural language:
```suggestion
const std::string db::INITIAL_RADIO_ENABLE_STR = "initial radio enable";
```

adam1985d replied:
the strings are used as parameter-key-names (aka options in UCI). although the BPL could have restructured key-names from/to the suggested strings I preferred not to introduce this translation - as there is no added value to it.

RanRegev commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2542,16 +2548,21 @@ bool db::set_client_time_life_delay(const sMacAddr &mac,
 
     auto timestamp = std::chrono::steady_clock::now();
     if (save_to_persistent_db) {
```
Saving or not saving to persistent is based on the truth status of two variables: `save_to_persistent_db && config.persistent_db`.
So maybe instead of asking these two questions separately each time, with a potential forget, maybe construct it a bit differently:
```cpp
// at the begin of each function - 
bool persistent_save = save_to_persistent_db && config.persistent_db;
/* ... */
if (persistent_save) { /* ... */ }
```

adam1985d replied:
I wish to indicate to the user that the request to save to the persistent DB was not met since the persistent DB is disabled. how about:
```
// at the begin of each function - 
bool persistent_save = save_to_persistent_db && config.persistent_db;
/* ... */
LOG_IF((save_to_persistent_db && !config.persistent_db), DEBUG) << "persistent db is disabled";
if (persistent_save) { /* ... */ }
```

RanRegev replied:
It is the same log each time you check, so I guess one time is enough.
I don't understand why you added the `LOG_IF` _after_ some code. (in your example: `/*...*/`)
For my understanding it should be immediately followed the assignment to `persistent_save` :
```cpp
bool persistent_save = save_to_persistent_db && config.persistent_db;
LOG(DEBUG) << "persistent_save is: " << std::boolalpha << persistent_save << " because save_to_persistent_db is " << save_to_persistent_db << " and config.persistent_db is " << config.persistent_db;
```
This message, for me, is the most useful one. I am not walking around the bush - I have all the information that I need in the log. Looking at the log and looking in the code I don't need to work hard to first translate the log, it is one to one with the code.
Moreover, note that I don't log only when some condition is met, I always log. This kind of logs are beacons in the log, indicating something in the form "was here and with this state". Note however that central places, that happens a lot are not good candidates for such logs.

adam1985d replied:
The reason is that when used properly - the API should execute silently and only provide debug logs for valuable information.
The only case where the flow doesn't fully perform what the caller is asking is the case where `save_to_persistent_db` is `true` but the persistent DB is actually disabled  (`config.persistent_db` is `false`) - all other cases are valid.
But I don't want to fail this flow - instead, provide the debug information and perform the runtime DB update.

RanRegev commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4227,3 +4227,28 @@ bool db::update_client_entry_in_persistent_db(
     return true;
 }
+
+bool db::add_client_entry_and_update_counter(
```
nitpick (but important):
functions that describes their _entire_ work tend to be with time either inaccurate or incomplete:
`db::add_client_entry_and_update_counter_and_report_and_take_care_for_errors`...
It is best to keep it simple: `add_client_entry`
the work that is done inside is irrelevant to the caller.

adam1985d replied:
well actually in this specific case it is very important - since I wish to distinguish that this private function manages the `m_persistent_db_clients_count` and any API calling it must not update the counter by itself.

RanRegev replied:
This is tightly related to my BLOB claim:
Once you have such a huge class with so many unrelated entities, tasks and concerns - you simply start to add signs, signals, hints to make sure other parts of the same thing are acting correctly. This is a bad sign.
The opposite would be that you have a small class, managing a small number of inter-related entities, and then this question, this concern, wouldn't rise at all. It is clear that among the 5-10 functions one is responsible to this member.

adam1985d replied:
This is true - and as mentioned in my other reply - there is a separate task for refactoring the db class for that purpose.

RanRegev commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -666,7 +666,7 @@ class db {
      */
     bool add_client_to_persistent_db(const sMacAddr &mac,
-                                     std::unordered_map<std::string, std::string> params =
+                                     const std::unordered_map<std::string, std::string> &params =
```
This is fine, though worth noticing:
when `params` is not supplied and the default is selected (after the = sign), `params` is bounded to a temporary object. It is a reference to temporary.
However, because of the `const` in the beginning, the lifetime of the temporary is extended to the lifetime of `params`.
https://godbolt.org/z/EEYaT7

adam1985d replied:
true, I wouldn't have used the reference and the default-value without the `const` :-)

RanRegev commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1104,6 +1104,16 @@ class db {
+     * @return true on success, otherwise false.
+     */
+    bool set_node_params_from_map(const sMacAddr &mac,
+                                  const std::unordered_map<std::string, std::string> &values_map);
```
I am mentioning it here, but could be anywhere this type is used.
Maybe it is a good idea to wrap `value_map` with at least `using` and at most its own class, instead of having it as a raw declaration.

at least -
```cpp
using value_map = std::unordered_map<std::string, std::string>;
```
at most -
```cpp
class ValueMap{
public:
// interfaces that allow setting only legal values
// checks consistency of values, validates and more...
private:
/*whatever suitable to hold this map*/
};
```

adam1985d replied:
True, I will add this as a separate commit in next PR for review clarity - as this improvement spreads across several APIs and API implementations in the db.h and db.cpp.

RanRegev commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1133,6 +1133,16 @@ class db {
+     *
+     * @return sMacAddr mac of candidate client to be removed - if not found, string_utils::ZERO_MAC is returned.
+     */
+    sMacAddr get_candidate_client_for_removal();
```
please see the comment in the implementation.

adam1985d replied:
answered you comment in the implementation.

RanRegev commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4302,3 +4302,67 @@ bool db::remove_client_entry_and_update_counter(const std::string &entry_name)
     return false;
 }
+
+sMacAddr db::get_candidate_client_for_removal()
```
This is application specific and have nothing to do with the DB.
Adding this functionality to the database mixes concerns in a way that makes it extremely hard to maintain from many aspects.
This function implements pure application's logic that **_uses_** information from the db.
To do: extract from the db class and move elsewere.

adam1985d replied:
this API is used by the DB for maintaining a seamless "add new client" on one hand - while keeping the "clients_persistent_db_max_size" limit on the other.

I agree this is pure logic - but since the db is the only one to use it and since the db encapsulates the managing of the size-limitation, I still think this should be part of the db class.
Moreover - the user **new** client configuration is always preferred over existing client configurations and will always be added even if removal of an entry from the DB needs to be performed to "make room" for it - so I don't want the user to handle it externally.

RanRegev replied:
See my later comment to the use of this functionality.
The db is responsible to remove-upon-adding-to-full, but the client is responsible to the select who to remove.
The db provides the mechanism, the rule, the action that will be taken - the "what", the client provides the "how".

adam1985d replied:
as discussed offline, I have decided to leave it as is at this stage - but will consider introducing this change if we really require more than one way of determining the candidate client to be removed.

RanRegev commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -1133,6 +1133,13 @@ class db {
+     * 
+     * @return true on success, otherwise false.
+     */
+    bool remove_candidate_client();
```
Belong to application's logic, not to db's logic.
must be removed elsewhere.

adam1985d replied:
lets continue discussing both `get_candidate_client_for_removal` and `remove_candidate_client` approach as part of https://github.com/prplfoundation/prplMesh/pull/1526/commits/06646be2a09cbd9ffd7316a9f07e3ccf1b7347a3#r457958443

RanRegev commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2513,6 +2513,17 @@ bool db::add_client_to_persistent_db(const sMacAddr &mac,
+    if (m_persistent_db_clients_count >= config.clients_persistent_db_max_size) {
+        LOG(DEBUG) << "reached max clients size in persistent db - removing a client before adding "
+                      "new client";
+        if (!remove_candidate_client()) {
```
if you want to _automatically_ remove "old" clients when requested to add a "new" client it is OK, but it doesn't mean that you need to _implement_ the way "old" is selected. You should leave it to the user, to the caller of `add_client_to_persistent_db`. The call to add a client is then interpreted as "add this client and if you need to remove someone, use this function to select who".
There are few ways to do so, here are two:
```cpp
<template old_client_selection_func>
bool db::add_client_to_persistent_db(const sMacAddr &mac,
                                     const std::unordered_map<std::string, std::string> &params,
                                     old_client_selection_func user_selection );
{
/* .... */
// and then
 if (!user_selection(*this)) {
 /* .... */ }
}

// note, old_client_selection_func has this interface:
bool selection(const db &);
```
Another, object oriented way:
```cpp
struct DbFunctions {
virtual bool select_client_to_remove(const db &) = 0;
virtual sMacAddr get_the_craziest_client(const db &) = 0;
/* etc */
};
// and let derived class implement their own logic.

// in the db class add this function:
void set_user_functionality(DbFunctions*);
DbFunctions* m_functionality = nullptr;

// and usage:
if (m_functionality && m_functionality->select_client_to_remove(*this) {
/* ... *.
```

adam1985d replied:
I suggest something in-between:
```
bool db::set_old_client_selection_func(old_client_selection_func user_selection_func) {
  m_user_selection_func = user_selection_func;
  return true;
}

Class db {
private:
  old_client_selection_func user_selection_func;
}
```

and then the set operation will be performed as part of the init flow of the controller.

RanRegev replied:
yes, as I said: _few_ ways. 
* The important issue - let the client supply the functionality of "how" to select the one to be removed.
* What is the type of `old_client_selection_func`?
* I prefer the general `user_selection` and attach it to each function that requires such a function. Because in the future the user might need to provide more functionality to solve more "how to" things for the database. Then the `user_selection` idiom is well understood and you don't need to invent a new member and a new interface each time such a need is introduced. But totally up to you.

adam1985d replied:
as discussed offline, I have decided to leave it as is at this stage - but will consider introducing this change if we really require more than one way of determining the candidate client to be removed.

RanRegev reviewed (changes requested):
a. Maybe I am too late, but this class became huge, a [BLOB](https://sourcemaking.com/antipatterns/the-blob) with mixed functionality: string utils (e.g. mac to string), pure get/set functions, and also application's specific logic (auto remove clients).
Maybe the best way to refactor it, to separates concerns, is to create a `namespace db` that lets us implement functionalities side by side under the db namespace but with a single role of responsibility for each of the concerns.
b. commit [5f64f98405](https://github.com/prplfoundation/prplMesh/pull/1526/commits/5f64f984050b2f5d5cb571e7037afb109c553ca3) states that 
> provides efficiency compared to performing

Is there any benchmark that proves this? Any unusable code before prior to this addition?
c. I didn't see any use of `m_persistent_db_clients_count` (introduced in [0d9917d2c0d](https://github.com/prplfoundation/prplMesh/pull/1526/commits/0d9917d2c0d5915565da2e606d6d68575dfce4e4) - but maybe I just missed it.

adam1985d commented:
> a. Maybe I am too late, but this class became huge, a [BLOB](https://sourcemaking.com/antipatterns/the-blob) with mixed functionality: string utils (e.g. mac to string), pure get/set functions, and also application's specific logic (auto remove clients).
> Maybe the best way to refactor it, to separates concerns, is to create a `namespace db` that lets us implement functionalities side by side under the db namespace but with a single role of responsibility for each of the concerns.
> b. commit [5f64f98405](https://github.com/prplfoundation/prplMesh/pull/1526/commits/5f64f984050b2f5d5cb571e7037afb109c553ca3) states that
> 
> > provides efficiency compared to performing
> 
> Is there any benchmark that proves this? Any unusable code before prior to this addition?
> c. I didn't see any use of `m_persistent_db_clients_count` (introduced in [0d9917d2c0d](https://github.com/prplfoundation/prplMesh/pull/1526/commits/0d9917d2c0d5915565da2e606d6d68575dfce4e4) - but maybe I just missed it.

Thank you for your comments.

There is a goal to refactor the DB but it is a separate task and is not addressed in this PR.
Regarding the m_persistent_db_clients_count - yes it is used and its purpose is to meet the configurable limitation of "clients_persistent_db_max_size"

RanRegev reviewed (approved):
The entire class should be refactored. It is planned as another task.
Other than that few minor issues.
